src/pin.rs

```diff
@@ -0,0 +1,663 @@
+//! enum List<'a, T> {
+//!     Cons(T, Box<'a, List<'a, T>>),
+//!     Nil,
+//! }
+//!
+//! let list: List<i32> = List::Cons(1, Box::new_in(List::Cons(2, Box::new_in(List::Nil, &b)), &b));
+//! println!("{:?}", list);
+//! ```
+//!
+//! This will print `Cons(1, Cons(2, Nil))`.
+//!
+//! Recursive structures must be boxed, because if the definition of `Cons`
+//! looked like this:
+//!
+//! ```compile_fail,E0072
+//! # enum List<T> {
+//! Cons(T, List<T>),
+//! # }
+//! ```
+//!
+//! It wouldn't work. This is because the size of a `List` depends on how many
+//! elements are in the list, and so we don't know how much memory to allocate
+//! for a `Cons`. By introducing a [`Box<'a, T>`], which has a defined size, we know how
+//! big `Cons` needs to be.
+//!
+//! # Memory layout
+//!
+//! For non-zero-sized values, a [`Box`] will use the provided [`Bump`] allocator for
+//! its allocation. It is valid to convert both ways between a [`Box`] and a
+//! pointer allocated with the [`Bump`] allocator, given that the
+//! [`Layout`] used with the allocator is correct for the type. More precisely,
+//! a `value: *mut T` that has been allocated with the [`Bump`] allocator
+//! with `Layout::for_value(&*value)` may be converted into a box using
+//! [`Box::<T>::from_raw(value)`]. Conversely, the memory backing a `value: *mut
+//! T` obtained from [`Box::<T>::into_raw`] will be deallocated by the
+//! [`Bump`] allocator with [`Layout::for_value(&*value)`].
+//!
+//! Note that roundtrip `Box::from_raw(Box::into_raw(b))` looses the lifetime bound to the
+//! [`Bump`] immutable borrow which guarantees that the allocator will not be reset
+//! and memory will not be freed.
+//!
+//! [dereferencing]: https://doc.rust-lang.org/std/ops/trait.Deref.html
+//! [`Box`]: struct.Box.html
+//! [`Box<'a, T>`]: struct.Box.html
+//! [`Box::<T>::from_raw(value)`]: struct.Box.html#method.from_raw
+//! [`Box::<T>::into_raw`]: struct.Box.html#method.into_raw
+//! [`Bump`]: ../struct.Bump.html
+//! [`Drop`]: https://doc.rust-lang.org/std/ops/trait.Drop.html
+//! [`Layout`]: https://doc.rust-lang.org/std/alloc/struct.Layout.html
+//! [`Layout::for_value(&*value)`]: https://doc.rust-lang.org/std/alloc/struct.Layout.html#method.for_value
+
+use core::marker::PhantomData;
+
+use {
+    crate::Bump,
+    {
+        core::{
+            any::Any,
+            borrow,
+            cmp::Ordering,
+            future::Future,
+            hash::{Hash, Hasher},
+            iter::FusedIterator,
+            mem::ManuallyDrop,
+            ops::{Deref, DerefMut},
+            pin::Pin,
+            task::{Context, Poll},
+        },
+        core_alloc::fmt,
+    },
+};
+
+/// An owned pointer to a bump-allocated `T` value, that runs `Drop`
+/// implementations.
+///
+/// See the [module-level documentation][crate::boxed] for more details.
+#[repr(transparent)]
+pub struct Box<'a, T: ?Sized> {
+    ptr: *mut T,
+    _marker: PhantomData<&'a mut T>,
+}
+
+impl<'a, T> Box<'a, T> {
```

Comment:
`T` must satisfy `T: 'a` (or be able to outlive `'a`) so that drop within the bump would not be called on invalid value.

```suggestion
impl<'a, T> Box<'a, T>
where
    T: 'a
{
```

Reply:
Actually `'a` is just an lifetime to an immutable reference of `Bump`, which can be subtyped into smaller lifetimes via covariance, so this is not good enough. Appropriate fix for would be to add a lifetime argument `'b` to `Bump` itself to restrict it and then restrict `T: 'b`.

Reply:
The marker `PhantomData<&'a mut T>` inside `Box<'a, T>` should produce the implied bound `T: 'a`, so your suggested change should be a noop.

> Actually 'a is just an lifetime to an immutable reference of Bump, which can be subtyped into smaller lifetimes via covariance, so this is not good enough.

I don't see how this is unsound. Even if you'd have `&'short Bump`, `T` must outlive `'short` (and `'long: 'short`). The returned ref to `T` will be borrowed for the shorter lifetime and can't live longer than that.

It would be great if you could provide an example where this unsoundness is demonstrated :)

Reply:
You drop `T: 'short` object within the `Bump: 'long`'s drop.

Reply:
Sorry, I don't quite understand what you are trying to say. It would really help the discussion if you could provide a quick poc of the unsoundness.

Aside, since this `pinned::Box` implementation is almost an 1:1 interface copy of the existing `boxed::Box` implementation, your claim also affects the current in-tree `Box` implementation. So again, I'd be really interested in seeing this unsoundness you are talking about.

Reply:
Ah I see the issue now. Yeh, it's actually very easy to trigger with something like

```rust
#[test]
fn pinned_failure() {
    struct Foo<'a>(&'a String);

    impl<'a> Drop for Foo<'a> {
        fn drop(&mut self) {
            println!("{}", self.0);
        }
    }

    let bump = Bump::new();

    let s = "hello".to_string();
    let p = Box::new_in(Foo(&s), &bump);
    mem::forget(p);
    drop(s);

    drop(bump);
}
```